R/chop.R

```diff
@@ -90,188 +90,253 @@ chop <- function(data, cols) {
-  # If only `NULL` values are in the row, the `NA` size is finalised to `0`.
   for (i in seq_len_width) {
     col <- x[[i]]
+    col_is_list <- x_is_list[[i]]
 
-    for (j in seq_len_size) {
-      # TODO: col[[j]] -> vec_slice2(col, j)
-      elt <- col[[j]]
+    if (!col_is_list) {
+      # Optimize rare non list-cols
+      x_sizes[[i]] <- vec_rep(1L, size)
+      x_nulls[[i]] <- vec_rep(FALSE, size)
+      next
+    }
 
-      old_size <- sizes[[j]]
-      new_size <- tidyr_size(elt)
+    info <- unchop_col_info(col, size, keep_empty)
 
-      sizes[[j]] <- tidyr_size2(old_size, new_size)
-    }
+    x[[i]] <- info$col
+    x_sizes[[i]] <- info$sizes
+    x_nulls[[i]] <- info$null
   }
 
-  sizes <- map_int(sizes, tidyr_size_finalise)
+  sizes <- reduce(x_sizes, unchop_sizes2)
 
-  has_ptype <- !is.null(ptype)
-  if (has_ptype && !is.data.frame(ptype)) {
-    abort("`ptype` must be a data frame")
-  }
+  info <- unchop_finalize(x, sizes, x_nulls, keep_empty)
+  x <- info$x
+  sizes <- info$sizes
 
-  # Initialize `cols` with ptypes to retain types when `x` has 0 size
-  cols <- map(x, df_unchop_ptype)
-  pieces <- vector("list", size)
+  out_loc <- vec_rep_each(seq_len_size, sizes)
 
-  names <- names(x)
-  names(cols) <- names
+  out_size <- sum(sizes)
+  out_cols <- vector("list", length = width)
 
   for (i in seq_len_width) {
     col <- x[[i]]
+    col_is_list <- x_is_list[[i]]
 
-    for (j in seq_len_size) {
-      # TODO: col[[j]] -> vec_slice2(col, j)
-      elt <- col[[j]]
-      size <- sizes[[j]]
-
-      # Recycle each row element to the common size of that row
-      pieces[[j]] <- tidyr_recycle(elt, size)
+    if (!col_is_list) {
+      out_cols[[i]] <- vec_slice(col, out_loc)
+      next
     }
 
-    if (has_ptype) {
-      col_ptype <- ptype[[names[[i]]]]
-    } else {
-      col_ptype <- NULL
-    }
+    col_name <- names[[i]]
+    col_ptype <- ptype[[col_name]] %||% attr(col, "ptype", exact = TRUE)
 
-    # After unchopping, all columns will have the same size
-    # thanks to recycling
-    col <- vec_unchop(pieces, ptype = col_ptype)
+    # Drop to a bare list to avoid dispatch
+    col <- unclass(col)
 
-    # Avoid `NULL` assignment, which removes elements from the list
-    if (is.null(col)) {
-      next
-    }
+    col_sizes <- x_sizes[[i]]
+    row_recycle <- col_sizes != sizes
+    col[row_recycle] <- map2(col[row_recycle], sizes[row_recycle], vec_recycle)
 
-    cols[[i]] <- col
-  }
+    col <- vec_unchop(col, ptype = col_ptype, name_spec = zap())
 
-  out_size <- sum(sizes)
+    if (is_null(col)) {
+      # This can happen when both of these are true:
+      # - `col` was an empty list(), or a list of all `NULL`s.
+      # - No ptype was specified for `col`, either by the user or by a list-of.
+      if (out_size != 0L) {
+        abort("Internal error: `NULL` column generated, but output size is not `0`.")
+      }
 
-  loc <- rep(seq_len_size, sizes)
+      col <- unspecified(0L)
+    }
```

Comment:
This edge case is mainly a side effect of the fact that `vec_unchop(list())` and `vec_unchop(list(NULL))` return `NULL`. There isn't much we can do about this, and I feel like it is the right vctrs behavior, but we can't insert a `NULL` as a data frame column so the correct alternative is to insert an unspecified object. It can come up when the user tries to unchop `tibble(x = list())` or `tibble(x = list(NULL))`, where there is clearly no obvious type for `x`.